src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+        /// Returns an enumerator over the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        public SequencePosition GetPosition(SequencePosition origin, long offset)
+        {
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+            return Seek(origin, BufferEnd, offset, false);
+        }
+
+        public bool TryGet(ref SequencePosition position, out ReadOnlyMemory<T> data, bool advance = true)
+        {
+            var result = TryGetBuffer(position, End, out data, out var next);
+            if (advance)
+            {
+                position = next;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// An enumerator over the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public struct Enumerator
+        {
+            private readonly ReadOnlyBuffer<T> _readOnlyBuffer;
+            private SequencePosition _next;
+            private ReadOnlyMemory<T> _currentMemory;
+
+            /// <summary>
+            ///
+            /// </summary>
+            public Enumerator(ReadOnlyBuffer<T> readOnlyBuffer)
+            {
+                _readOnlyBuffer = readOnlyBuffer;
+                _currentMemory = default;
+                _next = readOnlyBuffer.Start;
+            }
+
+            /// <summary>
+            /// The current <see cref="Memory{Byte}"/>
+            /// </summary>
+            public ReadOnlyMemory<T> Current => _currentMemory;
+
+            /// <summary>
+            /// Moves to the next <see cref="Memory{Byte}"/> in the <see cref="ReadOnlyBuffer{T}"/>
+            /// </summary>
+            /// <returns></returns>
+            public bool MoveNext()
+            {
+                if (_next.Segment == null)
+                {
+                    return false;
+                }
+
+                return _readOnlyBuffer.TryGet(ref _next, out _currentMemory);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ReadOnlyBuffer<T> SliceImpl(SequencePosition begin, SequencePosition end)
+        {
+            return new ReadOnlyBuffer<T>(
+                begin.Segment,
+                begin.Index & IndexBitMask | (Start.Index & ~IndexBitMask),
+                end.Segment,
+                end.Index & IndexBitMask | (End.Index & ~IndexBitMask)
+            );
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private BufferType GetBufferType()
+        {
+            // We take high order bits of two indexes index and move them
+            // to a first and second position to convert to BufferType
+            return (BufferType)((((uint)Start.Index >> 30) & 2) | (uint)End.Index >> 31);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
This shouldn't be necessary here.

Reply:
Why not?

Reply:
The JIT should be smart enough to definitely inline this automatically. If it doesn't, it would be a codegen issue. @AndyAyersMS, can you please confirm?